lib/AST/NameLookup.cpp

```diff
@@ -239,11 +239,25 @@ bool swift::removeShadowedDecls(SmallVectorImpl<ValueDecl*> &decls,
          firstIdx != n; ++firstIdx) {
       auto firstDecl = collidingDecls.second[firstIdx];
       auto firstModule = firstDecl->getModuleContext();
+      auto firstSig = firstDecl->getOverloadSignature();
       for (unsigned secondIdx = firstIdx + 1; secondIdx != n; ++secondIdx) {
         // Determine whether one module takes precedence over another.
         auto secondDecl = collidingDecls.second[secondIdx];
         auto secondModule = secondDecl->getModuleContext();
 
+        // Swift 4 compatibility hack: Don't shadow properties defined in
+        // extensions of generic types with properties defined elsewhere.
+        // This is due to the fact that in Swift 4, we only gave custom overload
+        // types to properties in extensions of generic types, otherwise we
+        // used the null type.
+        if (!ctx.isSwiftVersionAtLeast(5)) {
```

Comment:
What if we instead said that in Swift 5, all properties get a real overload signature? Why should properties be any different from methods in this regard?

Reply:
@slavapestov By "real overload signature", do you mean that two properties of different type would always be allowed to overload each other? I'm not mad on this idea, as IMO it would encourage tricky code that relies on overload resolution by "return type", which I believe is something we want to discourage (listed in the API design guidelines under "[General conventions](https://swift.org/documentation/api-design-guidelines/#general-conventions)" > "Methods can share a base name").

However given that we already allow what would otherwise be re-declarations across module boundaries for both properties and methods, I personally think that it seems reasonable to allow different typed properties to overload across modules (as the shadowing behaviour is otherwise a bit unintuitive).

Though whatever we choose, I agree we should definitely make the behaviour consistent in Swift 5 mode!